linkerd/protocol/http/src/main/scala/io/buoyant/linkerd/protocol/http/FramingFilter.scala

```diff
@@ -0,0 +1,91 @@
+package io.buoyant.linkerd.protocol.http
+
+import com.twitter.finagle._
+import com.twitter.finagle.buoyant.linkerd.Headers
+import com.twitter.finagle.http.{Status, _}
+import com.twitter.util.Future
+import com.twitter.logging.Logger
+import io.buoyant.linkerd.ProtocolException
+
+object FramingFilter {
+  // since the client and server filters go in separate stacks,
+  // they can have the same role.
+  val role = Stack.Role("FramingFilter")
+
+  // this is factored out so that the same logic can
+  // be applied to requests from clients and responses from services.
+  private[FramingFilter] def filterMessage[M <: Message](message: M): Future[M] = {
+    val headers = message.headerMap
+    val contentLengths = message.headerMap.getAll("Content-Length")
+
+    if (contentLengths.toSet.size > 1) {
+      // if the length of the Content-Length key in the request/response's
+      // header map is greater than 1, then there are duplicate values.
+      Future.exception(FramingException("conflicting `Content-Length` headers"))
+    } else {
+      if (contentLengths.nonEmpty &&
+          headers.get("Transfer-Encoding").contains("chunked")) {
+        // if the message contains both a `Content-Length` header and
+        // `Transfer-Encoding: chunked`, remove the `Content-Length` header
+        headers.remove("Content-Length")
+        // this *should* modify the headers in-place? i dislike mutability...
+      }
+      Future.value(message)
+    }
+
+  }
+
+
+  /**
+   * A filter that fails badly-framed requests.
+   */
+  val serverFilter = new SimpleFilter[Request, Response] {
+    // unlike the client-side filter, this needs its own logger, since
+    // it responds to bad requests directly, rather than bubbling up
+    // exceptions to an ErrorResponder
+    private[this] val log = Logger.get("FramingFilter.ServerFIlter")
+
+    override def apply(
+      request: Request,
+      service: Service[Request, Response]
+    ): Future[Response] =
+      filterMessage(request)
+        .flatMap(service(_))
+        .rescue { case e: FramingException =>
```

Comment:
As it's currently written, the `flatMap` has to be before the `rescue`/`handle`, since it changes the type from `Request` to `Response` – that's why changing it broke the build in 7dd76a64c7c1246b1243683a76a212322f7eacb4 (sorry).